xtask/src/main.rs

```diff
@@ -0,0 +1,477 @@
+    TestHost,
+    TestLint,
+    TestSnapshot,
+}
+
+fn run_command(cmd_and_args: &[&str], cwd: Option<&str>, env: &[(&str, &str)]) -> Result<()> {
+    let cmd_and_args = Vec::from(cmd_and_args);
+    let mut cmd = &mut Command::new(cmd_and_args[0]);
+    if cmd_and_args.len() > 1 {
+        cmd.args(&cmd_and_args[1..]);
+    }
+
+    for (k, v) in env {
+        cmd.env(k, v);
+    }
+
+    let cwd_s = if let Some(path_ref) = cwd {
+        cmd = cmd.current_dir(path_ref);
+        format!("CWD:{} ", path_ref)
+    } else {
+        "".to_string()
+    };
+
+    let cmdline = cmd_and_args.join(" ");
+    println!("🏃 {}{}", cwd_s, cmdline);
+    cmd.status()
+        .map_err(|e| anyhow!("could not run '{}{}': {}", cwd_s, cmdline, e))
+        .and_then(|exit_status| {
+            if exit_status.success() {
+                Ok(())
+            } else {
+                let info = match exit_status.code() {
+                    Some(code) => {
+                        format!("non-zero exit status: {}", code)
+                    }
+                    None => "killed by signal".to_string(),
+                };
+
+                Err(anyhow!("'{}' did not finish successfully: {}", cmdline, info))
+            }
+        })
+}
+
+fn run_capturing_stdout(cmd: &mut Command) -> Result<String> {
+    let child = cmd.stdout(Stdio::piped()).spawn()?;
+    let mut stdout = child
+        .stdout
+        .ok_or_else(|| anyhow!("could not access standard output"))?;
+    let mut out = String::new();
+    stdout
+        .read_to_string(&mut out)
+        .with_context(|| "could not read standard output")?;
+    Ok(out)
+}
+
+fn do_test<F>(t: F, context: &str)
+where
+    F: FnOnce() -> Result<()>,
+{
+    match t() {
+        Ok(_) => {}
+        Err(e) => ALL_ERRORS.lock().unwrap().push(format!("{}: {}", context, e)),
+    }
+}
+
+fn rustc_is_nightly() -> bool {
+    // if this crashes the system is not in a good state, so we'll not pretend to be able to recover
+    let out = run_capturing_stdout(Command::new("rustc").args(&["-V"])).unwrap();
+    out.contains("nightly")
+}
+
+fn load_expected_output(name: &str, release_mode: bool) -> Result<String> {
+    let path = Path::new("firmware/qemu/src/bin");
+
+    let filename;
+    if release_mode {
+        filename = format!("{}.release.out", name);
+    } else {
+        filename = format!("{}.out", name);
+    }
+
+    let path = path.join(filename);
+
+    // for error context closure
+    let path_str = path.to_str().unwrap_or("(non-Unicode path)").to_string();
+
+    let content = fs::read_to_string(path)
+        .with_context(|| format!("Failed to load expected output data from {}", path_str))?;
+    Ok(content)
+}
+
+fn test_single_snapshot(name: &str, features: &str, release_mode: bool) -> Result<()> {
+    let display_name = format!("{} ({})", name, if release_mode { "release" } else { "dev" });
+    println!("{}", display_name);
+    let cwd_name = "firmware/qemu".to_string();
```

Comment:
`.to_string()` shouldn't be needed here

```suggestion
    let cwd_name = "firmware/qemu";
```